site/content/blog/2021-04-23-optimizing-go-compiler-1.md

```diff
@@ -0,0 +1,227 @@
+---
+title: "Fixing a bug in the Go compiler as a newbie: a deep dive (I)"
+heading: "Fixing a bug in the Go compiler as a newbie: a deep dive (I)"
+description: ""
+summary: "Contributing to the Go compiler is within everyone's reach. The first part of this series of posts defines the issue we are going to investigate, and introduces the parts of the Go compiler involved in its fix. We will learn about Static Single Assignment, rewrite rules and some specific incantations to hack on the Go compiler."
+slug: optimizing-go-compiler-1
+date: 2021-04-23T12:00:00-05:00
+categories:
+    - "go"
+author: Alejandro García Montoro
+github: agarciamontoro
+community: alejandro.garcia
+---
+
+[bf48163e8f2b604f3b9e83951e331cd11edd8495](https://github.com/golang/go/commit/bf48163e8f2b604f3b9e83951e331cd11edd8495). That is the hash of one of the 47791 commits that, at the time of this writing, build the whole git history of the Go compiler. It is definitely not the most important commit there, and most of the other ones probably deserve way more attention. But I like `bf48163`. It is a good commit, I think. I may be a bit biased, of course, as it is my first contribution to Go.
+
+Not only that, it is my first contribution _ever_ to any compiler!
+
+What's really interesting here is that I am not an expert in Go---nor in compilers in general---at all. Sure, I know my way through the language to write software using it, but I had never thought I could be able to modify the compiler itself.
+
+And that is why I wrote this, to let you know that fixing a bug in a compiler is not only meant for a few chosen ones, but within reach of anyone.
+
+Although this was conceived as a single post, it got a wee bit out of hand, so I decided to split it in three:
+
+-   Part 1: what the bug looks like, and an introduction to some parts of the Go compiler.
+-   [Part 2](/blog/optimizing-go-compiler-2): what the bug actually is.
+-   [Part 3](/blog/optimizing-go-compiler-3): how to fix the bug.
+
+Do not get discouraged for the length of the content: it does not correlate with the complexity of the issue, but with the level of detail I tried to use when explaining it. I wanted to make sure I explained the ins and outs of how I approached the fix, and I do not assume any knowledge about compilers from the reader, so I tried to discuss everything in chapter and verse.
```

Comment:
There shouldn't be a period at the end of "explaining it" I think. Maybe place it at the end of "approached the fix". But I'll leave the grammatical stuff to Justine :P

Reply:
Hm, not sure. For context, this was the structure in my head:

Do not get discouraged for X: it is not because of Y, but because of Z. And Z happens because I had these intentions.

But yeah, I agree that Justine has the last word!

Reply:
Ah ok, I read it slightly differently.